src/rdb.c

```diff
@@ -1475,7 +1470,39 @@ int rdbSave(int req, char *filename, rdbSaveInfo *rsi) {
     if (fflush(fp)) { err_op = "fflush"; goto werr; }
     if (fsync(fileno(fp))) { err_op = "fsync"; goto werr; }
     if (fclose(fp)) { fp = NULL; err_op = "fclose"; goto werr; }
-    fp = NULL;
+
+    return C_OK;
+
+werr:
+    serverLog(LL_WARNING,"Write error while saving DB to the disk(%s): %s", err_op, strerror(errno));
+    if (fp) fclose(fp);
+    unlink(filename);
+    return C_ERR;
+}
+
+/* Save DB to the file. Similar to rdbSave() but this function won't use a
+ * temporary file and won't update the metrics. */
+int rdbSaveToFile(const char *filename) {
+    startSaving(RDBFLAGS_NONE);
+
+    if (rdbSaveInternal(SLAVE_REQ_NONE,filename,NULL) != C_OK) {
+        stopSaving(0);
+        return C_ERR;
+    }
+
+    stopSaving(1);
+    return C_OK;
+}
+
+/* Save the DB on disk. Return C_ERR on error, C_OK on success. */
+int rdbSave(int req, char *filename, rdbSaveInfo *rsi) {
+    char tmpfile[256];
+    char cwd[MAXPATHLEN]; /* Current working dir path for error messages. */
+
+    startSaving(RDBFLAGS_NONE);
+    snprintf(tmpfile,256,"temp-%d.rdb", (int) getpid());
+
+    if (rdbSaveInternal(req,tmpfile,rsi) != C_OK) goto werr;
```

Comment:
if rdbSaveInternal failed, we don't need to attempt to unlink the tempfile in the error handling.
same for when fsyncFileDir fails.
in both cases it'll just fail to unlink, so maybe we don't care since we're gonna ignore the error anyway.
WDYT?

Reply:
You mean we can leave that file there? 

Just to clean up temporary file, maybe we can separate parts. 
- We don't call unlink() if `rdbSaveInternal()` fails. (As it will call unlink() internally).
- We call `unlink()` if rename() or fsyncFileDir() fails.

Reply:
i mean that there's no temporary files.
if rdbSaveInternal failed, it already deletes it inside the function.
and if fsyncFileDir it fails after the rename, so the temporary file is already gone.

i think we should keep the unlink inside rdbSaveInternal, and either add a boolean flag to indicate when the unlinking is needed, or keep the unlink and add a comment that it may fail because the file is already deleted / renamed.

i'm not sure i care much about what happens i fsyncFileDir fails (as long as it prints an error).
not even sure we should report the error to the caller, but if we do and the file remains that's fine i think.

Reply:
Okay, now I see your point. 

Added a commit. Maybe we just call unlink() on rename failure only. 
If fsyncFileDir() fails, there is nothing to do anyway. We don't unlink().